docs.md

```diff
@@ -192,6 +192,12 @@ The system project cannot be deleted.
 
 Project quotas and default limits must be consistent with one another and must be sufficient for the requirements of active namespaces.
 
+#### Container default resource limit validation
+
+Validation mimics the upstream behavior of the Kubernetes API server when it validates LimitRanges.
+The container default resource configuration must have properly formatted quantities, with no negative values
+for any requests and limits. Limits for any resource must not be less than requests.
```

Comment:
I'm able to create limitranges with negative CPU just fine.

```yaml
apiVersion: v1
kind: LimitRange
metadata:
  name: cpu-resource-constraint
spec:
  limits:
  - default: # this section defines default limits
      cpu: "-500m"
    type: Container
```

Applying works:

```
$ k apply -f limitrange.yaml 
limitrange/cpu-resource-constraint created
```

Then I can view it:

```sh
$ k get limitrange -o yaml
apiVersion: v1
items:
- apiVersion: v1
  kind: LimitRange
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"v1","kind":"LimitRange","metadata":{"annotations":{},"name":"cpu-resource-constraint","namespace":"default"},"spec":{"limits":[{"default":{"cpu":"-500m"},"type":"Container"}]}}
    creationTimestamp: "2024-02-05T14:30:48Z"
    name: cpu-resource-constraint
    namespace: default
    resourceVersion: "1994455"
    uid: 79e4e545-4e3c-4bd1-8106-38c712ef8a2b
  spec:
    limits:
    - default:
        cpu: -500m
      defaultRequest:
        cpu: -500m
      type: Container
kind: List
metadata:
  resourceVersion: ""
```

I don't think there is a use for negative values though for CPU/memory so I think we should keep that check for better UI/UX, but the description seems wrong here that we're  mimicking the Kubernetes API validation.

Reply:
And now I'm wondering if someone is using project limits with negative values.. This fix would not break existing objects but it would break new objects with similar use case.

Difficult to know how impactful that would be.

CC: @MbolotSuse

Reply:
Good catch @tomleb. I think that we should not provide additional validation beyond what k8s provides. We don't want to make assumptions on what use cases are/are not relevant for end users. The description here was on purpose, we want to just make sure that we can create a valid limit range, the rest is up to the end user.

@maxsokolovsky can you review this and explain why validation was added for negative values?

Reply:
You're both correct.

```
apiVersion: v1
kind: LimitRange
metadata:
  name: cpu-resource-constraint
spec:
  limits:
  - defaultRequest:
      cpu: "-150"
    type: Container
```

The API server will accept this request. But when you create a regular pod without specifying resources and limits, you get an error.

```
$ k run nginx --image=nginx
The Pod "nginx" is invalid: spec.containers[0].resources.requests[cpu]: Invalid value: "-150": must be greater than or equal to 0
```

I originally thought that it'd be helpful for the user to alert about negative values early, as they aren't otherwise useful or meaningful in vanilla K8s LimitRange objects. But perhaps we should just stick to what the API server accepts now. And we could just rely on whatever internal errors from K8s might bubble up (if any).

Reply:
Interestingly, in my limited testing, I noticed that the API server rejects negative values on resource quotas.
For example:

```
apiVersion: v1
kind: ResourceQuota
metadata:
  name: pods-high
spec:
  hard:
    cpu: "-1000"
    memory: -200Gi
    pods: "-10"
```

Trying to submit the above manifest results in errors:
```
The ResourceQuota "pods-high" is invalid: 
* spec.hard[pods]: Invalid value: "-10": must be greater than or equal to 0
* spec.hard[cpu]: Invalid value: "-1k": must be greater than or equal to 0
* spec.hard[memory]: Invalid value: "-200Gi": must be greater than or equal to 0
```